### DISCUSSION TITLE
Crash in libnode.so due to conflicting timers

### DISCUSSION THREAD

**Discussion by Amanmishra1** [nodejs/help#4915] [Q&A]
### Version

v14.19.0

### Platform

arm64

### What steps will reproduce the bug?

Create a C++ application ( shared library ) using timers 
https://man7.org/linux/man-pages/man2/timer_create.2.html  
Load it inside a nodejs application. 
Real time signal 0 should occur

### How often does it reproduce? Is there a required condition?

Always

### What is the expected behavior?

No crash should occur

### What do you see instead?

Real time signal 0

### Additional information

We have a c++ library using timer_create API 
https://man7.org/linux/man-pages/man2/timer_create.2.html   ( it mentions it should be one per process )
the standalone application works perfect but when integrated with an UI (using nodejs ) Real time signal 0 occurs

The call stack suggests crash in libnode.so ( PFA screenshot )
![image](https://user-images.githubusercontent.com/42412832/158363708-b83d193f-4dd2-4a54-bf09-69d67fe1b66d.png)

Is there a way to disable timers in nodejs or you handle Timers in nodejs better so that C++ application using timers work fine within nodejs.

**Comment by targos** [comment]
@nodejs/libuv

**Comment by bnoordhuis** [comment]
I've converted this to a discussion because I don't think this is a node or libuv bug.

You can use timer_create() but you need to handle the signal by installing a signal handler, otherwise the process gets killed.

Aside: the line numbers in your screenshot are off, maybe because it's a release build.

**Reply by Amanmishra1** [reply]
Yes i added a signal handler in the code to ignore the signal that i get using SIG_IGN but this leads to failure of few test cases which I fail to understand. So I asked is there a way to disable timers inside libuv ? Or a better way i could write my signal handler ? 

//Code snippet
{
  struct sigaction new_action, old_action;

  new_action.sa_handler = handle_exception;
  sigemptyset(&new_action.sa_mask);
  new_action.sa_flags = SA_RESTART | SA_NOCLDSTOP;

  sigaction(SIGRTMIN, NULL, &old_action);
    if (old_action.sa_handler != SIG_IGN)
    {
        sigaction(SIGRTMIN, &new_action, NULL);
    }
  }
  
 void handle_exception(int signo)
{
 //Do nothing
}

**Reply by bnoordhuis** [reply]
Libuv timers and timer_create() timers have nothing to do with each other, they're completely different mechanisms.

I'd stay away from real-time signals, just use SIGALRM. The problem with SIGRTMIN is that all parts of the program need to agree on its value but build flags can change it.

**Reply by Amanmishra1** [reply]
> Libuv timers and timer_create() timers have nothing to do with each other, they're completely different mechanisms.
> 
> I'd stay away from real-time signals, just use SIGALRM. The problem with SIGRTMIN is that all parts of the program need to agree on its value but build flags can change it.

The problem is libuv (see the screenshot in 1st comment ) is throwing SIGRTMIN from libnode.so i am just handling the exception as shown in the previous comment
